Remove providing Exchange. Call Provide() from relevant places.

hsanjuan opened the pull request:
See https://github.com/ipfs/boxo/issues/963

Refactor the logic to perform Provides, when the component has been given a provider:
  * Remove `providing.Exchange`
  * Provide directly from Blockstore
  * Provide directly from pinner/merkledag on dag traversal
  * Provide from MFS whenever there is a call to `DAGService.Add`

As an alternative to providing from MFS, wrapping the DagService as a "providing DAG service", or building a DAGService backed by a providing blockstore, was considered. It was decided that MFS should carry a Provider because it matches how other modules do it and it's quite explicit, MFS calls Provide every time that there is a call to DAGService.Add, which seems enough.

hsanjuan reviewed (commented):
One problem I have is that of circular dependencies between the `provider` module and the `blockstore/pinner/merkledag`. The main issue is that the provide interfaces are not separate from the reprovider implementation. For the moment I have copied the interfaces into the several modules to avoid this.

hsanjuan commented:
More problems when trying to integrate in Kubo:
  * In order to have a provider, we need a blockstore, but in order to have a providing blockstore, we need a provider.
  * And so on...
  
  This means we need
    * something like blockstore.SetProvider(prov).
    * or perhaps in the blockstore case it would work with a blockstore wrapper + decorator.

- fx gets very complicated as we have providing strategies affecting initialization in multiple places.
- if we do setters, we have to update interfaces.

Need to think about it, potentially discuss in colo.

hsanjuan commented:
The `routing` libp2p package has also Provide interfaces

guillaumemichel reviewed (commented):
Nicely done!

> MFS is untouched, for I believe we need to wrap the DagService as a "providing DAG service", or build a DAGService backed by a providing blockstore. It seems these can be done externally though.

So it means that this PR will (momentarily) break providing for `mfs` and `pinned+mfs` reprovide strategies. IMO it is fine to address this elsewhere.

IIUC the different reprovide strategies would require:
* `all`: define `Provider()` on the blockstore
* `flat` define `Provider()` on the blockstore. Equivalent to `all` for providing (but different priority during reprovides).
* `roots`: define `WithRootsProvider()` on the pinner
* `pinned`: define `WithPinnedProvider()` on the pinner
* `pinned+mfs`: define `WithPinnedProvider()` on the pinner + TBD
* `mfs`: TBD

Is it correct?

hsanjuan commented:
> * `all`: define `Provider()` on the blockstore
> 
>     * `flat` define `Provider()` on the blockstore. Equivalent to `all` for providing (but different priority during reprovides).
> 
>     * `roots`: define `WithRootsProvider()` on the pinner
> 
>     * `pinned`: define `WithPinnedProvider()` on the pinner
> 
>     * `pinned+mfs`: define `WithPinnedProvider()` on the pinner + TBD
> 
>     * `mfs`: TBD
> 
> 
> Is it correct?

Yes

hsanjuan commented on pinning/pinner/dspinner/pin.go:
```diff
@@ -142,6 +169,10 @@ func New(ctx context.Context, dstore ds.Datastore, dserv ipld.DAGService) (*pinn
 	}
 
+	for _, o := range opts {
+		o.f(p)
```
I'll double-check

hsanjuan commented on ipld/merkledag/merkledag.go:
```diff
@@ -204,10 +205,12 @@ func FetchGraphWithDepthLimit(ctx context.Context, root cid.Cid, depthLim int, s
 		return false
 	}
 
+	opts = append(opts, Concurrent())
```
hmm yeah good catch. Concurrent() was a hardcoded option, but now we can override it.

hsanjuan commented:
Ok:

  * I have switched to libp2p routing interfaces (better than duplicating them for now)
  * I have added a SetKeyProvider function to the reprovider
  * I have moved NewPinnedProvider to dspinner and removed blockstore/pinner dependencies from the `provider` module.
  
  With this I can get away with my intentions and I can:
    * Create a provider.System without need for a blockstore/pinner in Kubo
    * Use the provider.System to set the Provider option for the blockstore/pinner
    * SetKeyProvider afterwards to the KeyChanFunc created using blockstore/pinner

guillaumemichel commented on provider/provider.go:
```diff
@@ -39,43 +34,26 @@ type System interface {
 // KeyChanFunc is function streaming CIDs to pass to content routing
 type KeyChanFunc func(context.Context) (<-chan cid.Cid, error)
 
-// NewBlockstoreProvider returns key provider using bstore.AllKeysChan
-func NewBlockstoreProvider(bstore blocks.Blockstore) KeyChanFunc {
-	return func(ctx context.Context) (<-chan cid.Cid, error) {
-		return bstore.AllKeysChan(ctx)
-	}
-}
```
Was this method moved elsewhere?

hsanjuan replied:
it was removed because == `bstore.AllKeysChan` to all effects (i don't even understand the indirection, probably it was a intermediate step before function signatures were aligned. Anyways, the blockstore dependency in this module needs to go away.

hsanjuan commented on exchange/providing/providing.go:
Yes, notifies blocks to bitswap so that bitswap knows it does not need to request them anymore, or that it can provide them if people requested them.

hsanjuan commented on blockstore/blockstore.go:
```diff
@@ -216,7 +230,15 @@ func (bs *blockstore) Put(ctx context.Context, block blocks.Block) error {
+
+	if bs.provider != nil {
+		logger.Debugf("blockstore: provide %s", block.Cid())
+		return bs.provider.Provide(ctx, block.Cid(), true)
```
Agreed on errors, disagree on not-blocking for Provide.

If you're provider is the reprovider, it just queues and returns. If you're provider is the (current) DHT, this call will effectively block for several seconds. I think it is ok that it does, because the alternative is launching goroutines (1000 blocks added -> 1000 concurrent provide operations goroutines). So for me the responsibility of handling concurrent provides is in the Provider implementation and the blockstore should not take the liberty to handle concurrency there.

guillaumemichel replied:
Ok, I agree with your reasoning 👍🏻

hsanjuan commented on blockstore/blockstore.go:
```diff
@@ -351,3 +381,21 @@ func (bs *gclocker) PinLock(_ context.Context) Unlocker {
+	}
+
+	for _, k := range keys {
+		logger.Debugf("reprovider: providing %s", k)
```
why? it's useful to know where provides are triggered from.

guillaumemichel replied:
If `boxo/provider:reprovider` is used, this log line is printed when the cid is added to the provide queue.

There is already [a log](https://github.com/ipfs/boxo/blob/2cf77c4c3ebf7898d3ff57a452ca1a33d6828a9c/provider/reprovider.go#L272) when a cid goes out of the provide queue to be passed to `kad-dht` for DHT provide.

It seems redundant to log when the cid is added to the provide queue, and when it is provided.

---

Furthermore, the text is confusing because it is the same that gets logged [when a key is reprovided](https://github.com/ipfs/boxo/blob/803408a01b73ae23f2a62521439e5828d93a3a7f/provider/reprovider.go#L576).

hsanjuan replied:
The log messages are prepended with the facility that they belong too, and the file:line that triggered them. So you would only see the messages if you have logging enabled on the reprovider, in particular. I think if is useful that if you enable logging on the blockstore, you get info on what the blockstore is doing even if that is redundant when logging is enabled for multiple things.

hsanjuan commented on provider/provider.go:
```diff
@@ -39,43 +34,26 @@ type System interface {
 	Clear() int
 	Close() error
 	Stat() (ReproviderStats, error)
+	SetKeyProvider(kp KeyChanFunc)
```
It will be called in Kubo, as I cannot create the blockstore without the reprovider, and I cannot reprovide from the blockstore otherwise. It's unfortunate.

lidel reviewed (commented):
@hsanjuan thoughts on opening a Kubo PR that uses this to see the impact + be a place where we can start adding end-to-end tests for `Reprovider.Strategy` in Kubo? 

Right now we lack test results that:
- confirm `Reprovider.Strategy=all|pinned|roots` work as expected
- show `Reprovider.Strategy=mfs|pinned+mfs` need to be fixed

lidel commented on blockstore/blockstore.go:
```diff
@@ -351,3 +381,21 @@ func (bs *gclocker) PinLock(_ context.Context) Unlocker {
 	return atomic.LoadInt32(&bs.gcreq) > 0
 }
+
+func doProvideManyHashes(ctx context.Context, r routing.ContentProviding, keys []multihash.Multihash) error {
```
nit:  reusable func that is used here and in  `provider/reprovider.go#doProvideMany`?

hsanjuan replied:
No, provideMany is only used here outside of reprovider. I don't want to import the reprovider system here. I don't think it causes dep-cycles anymore but relying on the ContentProvider interface only is cleaner.

lidel commented on ipld/merkledag/merkledag.go:
```diff
@@ -514,6 +538,21 @@ func parallelWalkDepth(ctx context.Context, getLinks GetLinks, root cid.Cid, vis
+					// Successfully fetched "ci".
+					// provide it when needed
+					if prov := options.Provider; prov != nil {
+						err := prov.Provide(ctx, ci, true)
+						if err != nil && options.ErrorHandler != nil {
+							err = options.ErrorHandler(root, err)
```
💭  should error handling show the actual failing ci, or the root one?
```suggestion
							err = options.ErrorHandler(ci, err)
```

hsanjuan replied:
Per Gui's suggestions, we are going to not abort operations on provide failures, so we are just going to log the failure (and it logs the CID in latest version of the code).

hsanjuan commented on pinning/pinner/dspinner/pin.go:
```diff
@@ -291,7 +335,14 @@ func (p *pinner) doPinDirect(ctx context.Context, c cid.Cid, name string) error
+	if err != nil {
+		return err
+	}
+	if p.rootsProvider != nil {
```
not necessary because there is no dag walking I think.

hsanjuan commented:
I have addressed comments and I have added a provider param to MFS constructors.

This breaks the API but I don't know if it's worth to avoid that. Seems pretty straightforward to fix (passing nil, or a provider).

In the end I have opted for carrying a Provider around in MFS because it matches how other modules do it and it's quite explicit, rather than doing a magic DAGService wrapper, or a mfs-only-providing-blockstore in Kubo. MFS calls Provide every time that there is a call to DAGService.Add, which seems enough.

hsanjuan commented:
Review: make sure godoc reflects that you can pass `nil` as the provider.

And fix the tests.

gammazero commented:
> Review: make sure godoc reflects that you can pass `nil` as the provider.
> 
> And fix the tests.

Done

hsanjuan commented:
Per my latest advancements with Kubo, I think we're good to merge this here for the moment.
